cub/test/catch2_test_device_for_each_in_extents.cu

```diff
@@ -0,0 +1,190 @@
+
+#  include <cub/device/device_for.cuh>
+
+#  include <thrust/detail/raw_pointer_cast.h>
+
+#  include <cuda/std/array>
+#  include <cuda/std/mdspan>
+#  include <cuda/std/span>
+
+#  include "c2h/catch2_test_helper.h"
+#  include "c2h/utility.h"
+#  include "catch2_test_launch_helper.h"
+
+// %PARAM% TEST_LAUNCH lid 0:1:2
+
+DECLARE_LAUNCH_WRAPPER(cub::DeviceFor::ForEachInExtents, device_for_each_in_extents);
+
+/***********************************************************************************************************************
+ * Host reference
+ **********************************************************************************************************************/
+
+template <int Rank = 0,
+          typename T,
+          typename ExtentType,
+          _CUB_TEMPLATE_REQUIRES(Rank == ExtentType::rank()),
+          typename... IndicesType>
+static void fill_linear_impl(c2h::host_vector<T>& vector, const ExtentType&, size_t& pos, IndicesType... indices)
+{
+  vector[pos++] = {indices...};
+}
+
+template <int Rank = 0,
+          typename T,
+          typename ExtentType,
+          _CUB_TEMPLATE_REQUIRES(Rank < ExtentType::rank()),
+          typename... IndicesType>
+static void fill_linear_impl(c2h::host_vector<T>& vector, const ExtentType& ext, size_t& pos, IndicesType... indices)
+{
+  using IndexType = typename ExtentType::index_type;
+  for (IndexType i = 0; i < ext.extent(Rank); ++i)
+  {
+    fill_linear_impl<Rank + 1>(vector, ext, pos, indices..., i);
+  }
+}
+
+template <typename T, typename IndexType, size_t... Extents>
+static void fill_linear(c2h::host_vector<T>& vector, const cuda::std::extents<IndexType, Extents...>& ext)
+{
+  size_t pos = 0;
+  fill_linear_impl(vector, ext, pos);
+}
+
+/***********************************************************************************************************************
+ * Function Objects
+ **********************************************************************************************************************/
+
+template <typename IndexType, int Size>
+struct LinearStore
+{
+  using data_t = cuda::std::array<IndexType, Size>;
+
+  cuda::std::span<data_t> d_output_raw;
+
+  template <typename... TArgs>
+  __device__ void operator()(IndexType idx, TArgs... args)
+  {
+    static_assert(sizeof...(TArgs) == Size, "wrong number of arguments");
+    d_output_raw[idx] = {args...};
+  }
+};
+
+/***********************************************************************************************************************
+ * TEST CASES
+ **********************************************************************************************************************/
+
+using index_types =
+  c2h::type_list<int8_t,
+                 uint8_t,
+                 int16_t,
+                 uint16_t,
+                 int32_t,
+                 uint32_t
+#  if CUB_IS_INT128_ENABLED
+                 ,
+                 int64_t,
+                 uint64_t
+#  endif
+                 >;
```

Comment:
Suggestion: I see that `int128` will cause a static assertion in `multiply_extract_higher_bits`, so I assume we don't support `ForEachInExtents` on `extents<int128, ...>`. Maybe we should document that explicitly or at least provide a good and early error message (maybe static assert right at the API entry point).